### DISCUSSION TITLE
GraphQL design: multilingual inputs and outputs

### DISCUSSION THREAD

**Discussion by mrtnzlml** [adeira/universe#1767] [Q&A]
I am trying to design multilingual input and output types and I am looking for some interesting design ideas.

My initial input kickoff:

```graphql
input ProductMultilingualInput {
  images: [ProductImageUploadable!]!
  price: ProductPriceInput!
  en_US: ProductMultilingualInputTranslations
  es_MX: ProductMultilingualInputTranslations
}

input ProductMultilingualInputTranslations {
  name: String!
  description: String!
}
```

Pros:

\+ I can submit EN only, ES only, or both together (that's what I want)

Cons:

\- The languages are not dynamic for now but strictly EN and ES. In the future, I can change it to an array of translations if I want to.

It can be done similarly for the output type (pseudo-code):

```graphql
type Product {
  id: ID!
  name: String!
  description: String!
  images: [String!]!
  translations: ProductTranslations
  # ...
}

type ProductTranslations {
  en_US: {
    name: String,
    description: String
  }
  # ...
}
```

Note: there is `name` on the product directly and `name` inside translations. This allows me to return the product already translated from DB for FE and also to return all the translations if needed (for editation purposes for example).

**Question**: how would you do it differently and why?

**Comment by michalsanger** [comment]
For the input I'd prefer dynamic input of translations from the beginning:
```graphql
input ProductMultilingualInput {
  images: [ProductImageUploadable!]!
  price: ProductPriceInput!
  translations: [ProductMultilingualInputTranslations!]!
}

input ProductMultilingualInputTranslations {
  locale: String! # maybe enum
  name: String!
  description: String!
}
```

**Reply by mrtnzlml** [reply]
I overcome my laziness and did it exactly like this. 😊 Now, the question is: do you see any potential issues with this design? 🤔

**Reply by michalsanger** [reply]
You will probably need some input validation logic in the mutation resolver, multiple inputs for a same locale could be a problem.
